Modules/Core/Common/include/itkImage.h

```diff
@@ -318,6 +318,51 @@ class ITK_TEMPLATE_EXPORT Image : public ImageBase<VImageDimension>
   unsigned int
   GetNumberOfComponentsPerPixel() const override;
 
+  /** Returns (image1 == image2). */
+  friend bool
+  operator==(const Image & lhs, const Image & rhs)
+  {
+    if ((lhs.GetBufferedRegion() != rhs.GetBufferedRegion()) || (lhs.m_Spacing != rhs.m_Spacing) ||
+        (lhs.m_Origin != rhs.m_Origin) || (lhs.m_Direction != rhs.m_Direction) ||
+        (lhs.m_InverseDirection != rhs.m_InverseDirection))
+    {
+      return false;
```

Comment:
Should `operator==` also check if the _requested_ region is the same for both images? I mean, is the "requested region" a fundamental part of the _value_ of an image? Or can two images be considered equal when only their _requested_ region is different? What do you think?

Reply:
Since we are comparing all the other attributes, we should compare requested region too.

Reply:
OK, thanks @dzenanz. I was just thinking of `std::string`. Its `operator==` does not compare the capacity of the strings. Two strings with different capacity may still compare equal, because the capacity is not a fundamental part of the _value_ of a string.

But I'll add requested region to the comparison, if you think that's what we should do.

Reply:
My opinion is not strong, so let's hear @blowekamp or maybe @hjmjohnson.

Reply:
My initial aim is being able to test the new image functions as follows:

    itk::WriteImage(image1, fileName);
    const auto image2 = itk::ReadImage(file)
    ASSERT_EQ(*image2, *image1); // Asserts that *image2 == *image1

In that case the requested region is equal to the buffered region, as far as I know. (Right?)

However I just realized, the requested region would matter when testing `Image::Graft` as follows:

    image2->Graft(image1);
    ASSERT_EQ(*image2, *image1);

`ImageBase::Graft` copies both the buffered and the requested region. So OK, I'll amend the PR! 😃

Reply:
Update: The latest [force-push](https://github.com/InsightSoftwareConsortium/ITK/compare/20f6b9672f272ae55270d79ece92748e2257a8b2..9bd662bf7e3a6346b0099660ba27aa2cab9b40de) has included the requested region with the comparison  😃

Reply:
@dzenanz Honestly I'm still not 100% sure if the requested region should be included with the comparison. Especially now that I see that `itk::WriteImage(inputImage, fileName)` does adjust the requested region of the input image! Internally, via `ImageFileWriter::Write()`. And you would expect that an image _after_ calling `WriteImage` is "equal" to the image _before_ calling `WriteImage`. Right?

So then, would it be a bug of `ImageFileWriter::Write()` that is modifies the requested region?

Reply:
I would consider the requested region a "pipeline state" variable similar to the m_Modified time. We certainly don't want to compare the latter :)

I would be more inclined to ensure that the largest possible region was the same.

Reply:
OK, thanks Brad, I just removed requested region again, with the latest [force-push](https://github.com/InsightSoftwareConsortium/ITK/compare/18bc9596f4eba93efb5e85e8b268fcb962840f32..895645b296fe30579acce97c38665aca4584e3c3). Feeling like a flip flop  ;-)  Users who need to include the requested region with their comparison must now do so explicitly:

    if ((image1 == image2) && (image1.GetRequestedRegion() == image2.GetRequestedRegion())) // ...
	
Or an extra ASSERT_EQ, in unit tests:

    ASSERT_EQ(image1, image2);
    ASSERT_EQ(image1.GetRequestedRegion(), image2.GetRequestedRegion());